### PR TITLE
fix(rpc): disable balance check if call gas price is 0

### PR DESCRIPTION
**Pull request by mattsse**
this fixes the insufficient fund error as seen here #3721

we were missing a `disable_balance_check` in revm that happens when evm starts

```rust
        // Check if account has enough balance for gas_limit*gas_price and value transfer.
        // Transfer will be done inside `*_inner` functions.
        if balance_check > *caller_balance && !disable_balance_check {
            return Err(InvalidTransaction::LackOfFundForGasLimit {
                gas_limit: balance_check,
                balance: *caller_balance,
            }
            .into());
        }
```

this is now disabled if gas price is 0, which now behaves like other clients

### CODE REVIEW

File: crates/rpc/rpc/src/eth/api/transactions.rs [hunk 1]
```diff
@@ -543,6 +543,8 @@ where
         let mut db = SubState::new(State::new(state));
 
         let env = prepare_call_env(cfg, block_env, request, &mut db, overrides)?;
+        dbg!(&env.cfg);
+        dbg!(&env.block);
```
[the remark this answers is not in the document]
Comment (mattsse) [reply crates/rpc/rpc/src/eth/api/transactions.rs:547]: -.-

**Comment by shekhirin** [comment]
can you point to a place in Geth code where this check happens? I only found https://github.com/ethereum/go-ethereum/blob/e1fe6bc8469c626afaa86b1dfb819737e980a574/core/state_transition.go#L230-L250

**Comment by mattsse** [comment]
I'm having a hard time reading this, but you can try

```
{
    "jsonrpc": "2.0",
    "method": "eth_call",
    "params": [
        {
            "from": "0x0000000000000000000000000000000200000000",
            "to": "0x6b175474e89094c44da98b954eedeac495271d0f",
            "gasPrice": "0x0", 
            "data": "0x70a082310000000000000000000000006E0d01A76C3Cf4288372a29124A26D4353EE51BE"
        },
        "latest"
    ],
    "id": 1
}
```

**Comment by mattsse** [comment]
moving this to draft because this apparently only should be applied to `trace_call`:

https://github.com/ledgerwatch/erigon/blob/ef4ccd47736019132c1967d6921f6c4b014eb515/core/state_transition.go#L182-L183

File: crates/rpc/rpc/src/eth/revm_utils.rs [hunk 2]
```diff
@@ -227,6 +227,12 @@ where
 
     let mut env = build_call_evm_env(cfg, block, request)?;
 
+    if env.tx.gas_price == U256::ZERO {
+        // if gas price is not specific or was set to 0, we disable the balance check, which
+        // bypasses the initial `LackOfFundForGasLimit` check
+        env.cfg.disable_balance_check = true;
+    }
+
```
Comment (gakonst) [inline crates/rpc/rpc/src/eth/revm_utils.rs:235]: Do we want to go with the Erigon way here? This seems like a very weird thing to do. @snoopymev pls advise.

File: crates/rpc/rpc/src/eth/revm_utils.rs [hunk 3]
```diff
@@ -227,6 +227,12 @@ where
 
     let mut env = build_call_evm_env(cfg, block, request)?;
 
+    if env.tx.gas_price == U256::ZERO {
```
Comment (gakonst) [inline crates/rpc/rpc/src/eth/revm_utils.rs:230]: Do we want to go with the Erigon way here? This seems like a very weird thing to do.

**Pull request closed without merging**